### ISSUE TITLE
NullValuePropertyMappingStrategy.SET_TO_DEFAULT with default values

### ISSUE THREAD

**Issue by luisalves00** [mapstruct/mapstruct#1909]
I was trying to map a null value (source = "print" is null) to an object with some default values (e.g.  defaultValue = "A4"). Came up with:

` @Mapping(target = "print", source = "print", defaultExpression = "java(printSpecificationFrom(new me.app.PrintSpecificationDto()))")
    Project parcelFrom(ProjectDto source);

    @Mapping(target = "size", source = "size", defaultValue = "A4")
    @Mapping(target = "format", source = "format", defaultValue = "Landascape")
    @Mapping(target = "version", source = "version", defaultValue = "1")
    PrintSpecification printSpecificationFrom(PrintSpecificationDto source);`

because:

`   @Mapping(target = "print", source = "print", nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.SET_TO_DEFAULT)
    Project parcelFrom(ProjectDto source);

    @Mapping(target = "size", source = "size", defaultValue = "A4")
    @Mapping(target = "format", source = "format", defaultValue = "Landascape")
    @Mapping(target = "version", source = "version", defaultValue = "1")
    PrintSpecification printSpecificationFrom(PrintSpecificationDto source);`

is not calling printSpecificationFrom to set the default values, i.e. It returns new PrintSpecification() without setting the defaults.  

I'm not sure if it's a bug or I'm not using the nullValuePropertyMappingStrategy  correctly

**Comment by Captain1653** [comment]
@luisalves00 It's little hard to say, without you Source and Target, but I suppose you want something like that? If I'm right, then I'll publish my example on my Github.
![Снимок экрана от 2019-09-20 22-17-21](https://user-images.githubusercontent.com/4740207/65352907-7bfa1180-dbf4-11e9-8c96-77f77674a893.png)

**Comment by luisalves00** [comment]
@Captain1653 Exactly...but I have the exact same solution as you. Maybe I didn't explain it correctly.
Would be nice that Mapstruct had a feature to avoid:
`defaultExpression = "java(printSpecificationFrom(new me.app.PrintSpecificationDto()))")`
as with the:
`NullValuePropertyMappingStrategy.SET_TO_DEFAULT`
it almost generate the proper code.

**Comment by sjaakd** [comment]
the `NullValuePropertyMapping` is only effective in update mappings. That is not entirely clear from its name (and I still regret that because I keep on answering questions about that one on here & on SO). 

@filiphr .. I really would prefer deprecating this name and replacing it by a more appropriate  name in 2.0. Alternatively we can extend its use to regular mappings (although it will conflict with `NullValueCheck` strategy..

**Comment by sjaakd** [comment]
@luisalves00 .. its probably not the answer you hoped for, but it works as designed current.y.

**Comment by luisalves00** [comment]
@sjaakd I was just requesting an enhancement, as when I looked at the generated code when using the NullValuePropertyMapping it almost did what is need and would avoid the use of the "complex" expression. Maybe with little effort you can do it for 2.0, as currently the use case can be done using the defaultExpression.

**Comment by sjaakd** [comment]
@luisalves00 , but I think you are right. Partly the naming of `NullValuePropertyMapping` is to blame for the it, but also partly because the `SET_TO_DEFAULT` case is not supported in regular mappings in any other way. Note that the other options in `NullValuePropertyMapping` do not make sense in a regular mapping method, but only in an update mapping. 

 So I guess there are two options forward: 
1. apply `NullValuePropertyMapping` to regular mappings, but reconsider `NullValueCheck` strategy
2. renaming `NullValuePropertyMapping` to `NullValueMappingTargetPropertyMappingStrategy` or `NullValuePropertyMergeMappingStrategy` or something like that.. 

anyway.. the current API causes confusion.

**Comment by sjaakd** [comment]
I'll re-open this issue as feature request.

**Comment by luisalves00** [comment]
Great. Thanks :)
